llvm/include/llvm/ADT/PagedVector.h

```diff
@@ -0,0 +1,301 @@
+  // the elements however depends on the order of access of the pages.
+  uintptr_t Allocator = 0;
+
+  constexpr static T *invalidPage() { return reinterpret_cast<T *>(SIZE_MAX); }
+
+public:
+  // Default constructor. We build our own allocator.
+  PagedVector()
+      : Allocator(reinterpret_cast<uintptr_t>(new BumpPtrAllocator) | 0x1) {}
+  PagedVector(BumpPtrAllocator *A)
+      : Allocator(reinterpret_cast<uintptr_t>(A)) {}
+
+  ~PagedVector() {
+    // If we own the allocator, delete it.
+    if (Allocator & 0x1) {
+      delete getAllocator();
+    }
+  }
+
+  // Get the allocator.
+  BumpPtrAllocator *getAllocator() const {
+    return reinterpret_cast<BumpPtrAllocator *>(Allocator & ~0x1);
+  }
+  // Lookup an element at position Index.
+  T &operator[](std::size_t Index) const { return at(Index); }
+
+  // Lookup an element at position i.
+  // If the associated page is not filled, it will be filled with default
+  // constructed elements. If the associated page is filled, return the element.
+  T &at(std::size_t Index) const {
+    assert(Index < Size);
+    assert(Index / PAGE_SIZE < PageToDataIdx.size());
+    auto *&PagePtr = PageToDataIdx[Index / PAGE_SIZE];
+    // If the page was not yet allocated, allocate it.
+    if (PagePtr == invalidPage()) {
+      PagePtr = getAllocator()->template Allocate<T>(PAGE_SIZE);
+      // We need to invoke the default constructor on all the elements of the
+      // page.
+      for (std::size_t I = 0; I < PAGE_SIZE; ++I) {
+        new (PagePtr + I) T();
+      }
+    }
+    // Dereference the element in the page.
+    return *((Index % PAGE_SIZE) + PagePtr);
+  }
+
+  // Return the capacity of the vector. I.e. the maximum size it can be expanded
+  // to with the expand method without allocating more pages.
+  std::size_t capacity() const { return PageToDataIdx.size() * PAGE_SIZE; }
+
+  // Return the size of the vector. I.e. the maximum index that can be
+  // accessed, i.e. the maximum value which was used as argument of the
+  // expand method.
+  std::size_t size() const { return Size; }
+
+  // Expands the vector to the given NewSize number of elements.
+  // If the vector was smaller, allocates new pages as needed.
+  // It should be called only with NewSize >= Size.
+  void expand(std::size_t NewSize) {
+    // You cannot shrink the vector, otherwise
+    // one would have to invalidate contents which is expensive and
+    // while giving the false hope that the resize is cheap.
+    if (NewSize <= Size) {
+      return;
+    }
+    // If the capacity is enough, just update the size and continue
+    // with the currently allocated pages.
+    if (NewSize <= capacity()) {
+      Size = NewSize;
+      return;
+    }
+    // The number of pages to allocate. The Remainder is calculated
+    // for the case in which the NewSize is not a multiple of PAGE_SIZE.
+    // In that case we need one more page.
+    auto Pages = NewSize / PAGE_SIZE;
+    auto Remainder = NewSize % PAGE_SIZE;
```

Comment:
Could you spell out these types? If you look at the surrounding lines in isolation, the types are not immediately obvious.